File: Leetcode - Java/694_graph_numDistinctIslands.java
// https://leetcode.com/problems/number-of-distinct-islands/

class Solution {
    public int numDistinctIslands(int[][] grid) {
        Set<String> set = new HashSet();
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[i].length; j++) {
                if (grid[i][j] == 1) {
                    StringBuilder sb = new StringBuilder();
                    dfs(grid, i, j, i , j, sb);
                    set.add(sb.toString());
                }
            }
        }
        return set.size();
    }
    
    private void dfs(int[][] grid, int i, int j, int r, int c, StringBuilder sb) {
        int m = grid.length, n = grid[0].length;
        if (r >= 0 && r < m && c >= 0 && c < n && grid[r][c] == 1) {
            grid[r][c] = 0; 
            sb.append(String.valueOf(r - i) + String.valueOf(c - j));
            dfs(grid, i, j, r-1, c, sb);
            dfs(grid, i, j, r+1, c, sb);
            dfs(grid, i, j, r, c-1, sb);
            dfs(grid, i, j, r, c+1, sb);
        }
    }
}

File: Leetcode - Python/694_graph_numDistinctIslands.py
# https://leetcode.com/problems/number-of-distinct-islands/

class Solution:
    def numDistinctIslands(self, grid: List[List[int]]) -> int:
        islands = set()

        for i in range(len(grid)):
            for j in range(len(grid[i])):
                island = []
                front = [(i, j)]

                while front:
                    nxt = []
                    for x, y in front:
                        if 0 <= x < len(grid) and 0 <= y < len(grid[x]) and grid[x][y] == 1:
                            grid[x][y] = 0
                            island.append((x - i, y - j))
                            for m, n in (x + 1, y), (x - 1, y), (x, y + 1), (x, y - 1):
                                nxt.append((m, n))
                    front = nxt

                if island and str(island) not in islands:
                    islands.add(str(island))

        return len(islands)
